### DISCUSSION TITLE
Why use listeners and not just `onChange`?

### DISCUSSION THREAD

**Discussion by bennettdams** [TanStack/form#1046] [General] [answered]
Hey!

I would like to know about the intended use case of the "listeners" API.
The [docs](https://tanstack.com/form/latest/docs/framework/react/guides/listeners) talk about "Side effects for event triggers", where you can execute something of the field changes, e.g.

*"Country" input changes, "Province" field is resetted*

The example has a listener on the "Country" field:
```tsx
      <form.Field
        name="country"
        listeners={{
          onChange: ({ value }) => {
            console.log(`Country changed to: ${value}, resetting province`);
            form.setFieldValue("province", "");
          },
        }}
      >
        {(field) => (
          <label>
            <div>Country</div>
            <input
              value={field.state.value}
              onChange={(e) => field.handleChange(e.target.value)}
            />
          </label>
        )}
      </form.Field>
```

But why not just execute this update in the `onChange` function (at the `<input />` element) of the "Country" field? Something like:

```tsx
      <input
        value={field.state.value}
        onChange={e => {
          field.handleChange(e.target.value)
          form.setFieldValue("province", "")
        }}
      />
```

Is it just to have a cleaner separation of concerns or are the other benefits in using the "listener"?
Maybe the "listener" is executed **after** a succesful validation?
If so, I think it might be worth to add the difference between the two in the docs.

**Comment by sohamnandi77** [comment]
I feel listeners API is more useful in cases where we want to trigger event handlers of field based on other field's value change.

**Comment by juanvilladev** [comment]
I think it separates the concerns from the basic onChange updating the form with the new value and any side effects that should occur. On our project that contains complex validations it's a lifesaver.

**Comment by harry-whorlow** [comment, marked as answer]
It's really cool seeing people discuss the api 🤟

Well as you mentioned the separation of concerns, but also because onChange isn't the only event you can subscribe to. 

- onChange
- onBlur
- onMount
- onSubmit 

All those events can be subscribed to in the listeners api, could you imagine what your component would look like trying to subscribe to these events... And as everyone else has mentioned, handleChange should handle the change of the field value, its nice to be able to separate out the logic when you have dependent fields.

It also allows you to abstract away your components from the form field logic, so your NumberInput component doesn’t need to account for how it relates to other form fields, thats now the job of Field. 

this
```tsx
<Field
  name="age"
  children={(handleChange, state, ...// rest of props ) => (
    <NumberInput 
      state={state} 
      handleChange={(e) => {
        field.handleChange(e.target.value)
        form.setFieldValue("permissions", [])
      }}
      // rest of props
    />
  )}
/>
```

can become
```tsx
<Field
  name="age"
  listeners={{
    onChange: ({ value }) => {
      setFieldValue('permissions', []);
    },
  }}
  children={(props) => (
    <NumberInput props={props} />
  )}
/>
```

But, that's more subjective, personally I prefer it.

Hope this helps explain the reasoning!

**Reply by bennettdams** [reply]
That makes a lot of sense, thank you! Maybe its worth adding such intentions and use cases to the docs?

**Reply by harry-whorlow** [reply]
No worries, happy to help! 

Yeah it could be improved, however I think the maintainers have a lot on their plate ATM. Perhaps, at a future date a PR could be opened, I just don't want to bother them too much right now.

**Reply by chris-tse** [reply]
Came to the discussions with the exact question OP had in mind.

Just to check my understanding:
- `listeners.onChange` adds additional logic to run on top of `field.handleChange()`
- `field.handleChange()` triggers anything that is subscribed to the field via either `useStore` or `form.Subscribe`

Would it be accurate to say that `field.handleChange` could be described as `broadcastChangeToSubscribers`? Meaning you can conditionally trigger whether you want just a local `onChange` to run, or if you want to actually let subscribers know whether it changed? I'm currently using it as such:

```ts
const [isValidValue, setIsValidValue] = useState(false)

const myValue = useStore(form.store, (state) => state.values.myValue)

// tRPC with tanstack query
// fetch using input value if valid on change
const data = api.data.fetchata.useQuery(
  { value: myValue },
  {
    enabled: isValidValue && !!myValue,
  },
)



function handleInputChange(value: string, broadcast: (value: string) => void) {
  // early return on some conditions
  if (value.length < 5) return 

  const isValidValue = someValidation(value)

  if (isValidValue) {
    setIsValidValue(true)
    broadcast(value)
  }
}

// ...


<input
  name={field.name}
  onChange={e => handleInputChange(e.target.value, field.handleChange)}
/>

```

This seems to bypass / require duplication of validation via the `validators` prop but allows more fine grained control of subscriptions.

**Reply by harry-whorlow** [reply]
@chris-tse this is the code of the setValue which is wrapped by handleChange. so yes, onChange listener adds additional functionality on top of the relevant event.

<img width="655" alt="Screenshot 2025-03-18 at 08 45 40" src="https://github.com/user-attachments/assets/d834a064-2b05-487c-b0bc-7fc4be116852" />

...and the implementation is similar for the other events you can subscribe to. The idea of listeners is to provide you with the formApi at the form.Field level, and allow you to abstract away logic that is concerning other fields out of the handle change. 

If you start abstracting away the components from the form as done in [form composition](https://tanstack.com/form/latest/docs/framework/react/guides/form-composition), It starts to make a bit more sense. 

With regards to your code snippet, personally if it was me I would have a validator onBlur for something like `if (val > 5) return 'needs to be longer than 5'`, to me it looks like you've re-implemented the logic we've wrote, but without additional context I cant really say.  🤟

I hope this helps

**Reply by luchillo17** [reply]
I know this is old, still...

Another case I reach a lot for the listeners very often is context bound fields, where the `onChange` prop is not exposed, for the most part if I have a bound field using context I see no reason to override the `onChange` handler, and the listener I can use in case I need to change another field based on this one (cascading change for dependent fields).

```tsx
    <form.AppField
          name="modality"
          listeners={{
            onChange: () => {
              form.setFieldValue('subModality', '');
              form.setFieldValue('variantKey', '');
            },
          }}
        >
          {(field) => (
            <field.ComboboxField
              label={td('serviceParams.modality')}
              options={[
                {
                  value: 'intramural',
                  label: tForms('intramural'),
                },
                {
                  value: 'extramural',
                  label: tForms('extramural'),
                },
              ]}
            />
          )}
        </form.AppField>
```
In this case the `ComboboxField` automatically binds itself to the field context, I only pass the label and the options, and then in the listeners I can be specific about only managing the other fields that should clear up when this one changes.

**Reply by luchillo17** [reply]
If anything the only thing I wish I could do is have the previous value to check if the value actually changed, the listeners trigger even if the value is the same (when I click the already selected item).
